### PR TITLE
Support contracts

### LINKED ISSUE #43: Allow coverage to be run on contracts
In Cairo, invoking contracts creates separate VM instances, and this information is stored under the `nested_calls` key in trace files. Currently, the coverage computation does not analyze these nested calls, which prevents complete coverage analysis. The goal of this task is to collect all nested call traces so that coverage can be accurately computed including contracts.

```json
{
  "entry_point": {},
  "used_execution_resources": [],
  "used_l1_resources": [],
  "nested_calls": [], <-- here
  "cairo_execution_info": {}
}
```

**Implementation Steps:**

1. Create an empty hash set (`trace_file_set`) to store unique trace files encountered, including nested calls.

2. Implement a DFS algorithm to traverse the trace files. Recursively explore the `nested_calls` for each trace file node.

3. For each trace file node encountered during DFS, add it to the `trace_file_set`.

### PR DESCRIPTION
**Pull request by ksew1**
Closes #43

### CODE REVIEW

File: crates/cairo-coverage/src/data_loader/loaded_data.rs [hunk 1]
```diff
@@ -1,65 +1,84 @@
...
-                let VersionedProgram::V1 {
-                    program:
-                        ProgramArtifact {
-                            program,
+        Ok(Self(
+            execution_infos_by_sierra_path
+                .into_iter()
+                .map(|(source_sierra_path, casm_level_infos)| {
+                    read_and_deserialize::<SierraProgram>(&source_sierra_path)?
+                        .compile_and_get_debug_infos()
```
Comment (THenry14) [inline crates/cairo-coverage/src/data_loader/loaded_data.rs:50]: general question: how does this affect the execution times? 
I wonder if there is anything we can do beforehand to have this or at least not to compile each execution (and - should we). Please let me know if this is something that was already discussed/though of beforehand.

File: crates/cairo-coverage/src/main.rs [hunk 2]
```diff
@@ -2,40 +2,106 @@ mod cli;
...
+    project_path: &Option<Utf8PathBuf>,
+) -> Result<Utf8PathBuf> {
+    if let Some(project_path) = project_path {
+        Ok(project_path.clone())
+    } else {
+        find_user_project_path(source_sierra_path).context(indoc! {
+            r"Inference of project path failed.
+            Please provide the project path explicitly using the --project-path flag.
+            If you are using snforge, it is not possible to use cairo-coverage flags.
+            You need to run cairo-coverage on your own."
```
Comment (THenry14) [inline crates/cairo-coverage/src/main.rs:69]:
```suggestion
            You need to run cairo-coverage directly."
```
nit

File: crates/cairo-coverage/src/data_loader/loaded_data.rs (hunk 1 above)
Comment (ksew1) [reply crates/cairo-coverage/src/data_loader/loaded_data.rs:50]: I know compiling in a loop might seem intimidating, but:

- We group the files in a hashmap beforehand so we only compile unique ones. This is a minimal number as each program needs to be compiled to CASM to get its debug info.
- I tested this on the OpenZeppelin repo, and it doesn't seem like a bottleneck (I know it is not a reliable benchmark).


The same strategy is used in the profiler, and from what I've heard, profiling is fast enough. So, these two projects have the same issue. We can try to work on it together if it seems like a major performance issue.

Comment (ksew1) [reply crates/cairo-coverage/src/data_loader/loaded_data.rs:50]: The only edge case I can think of where we need to compile more than necessary is when `ContractClass` and `VersionedProgram` are the same program. But I haven't found a reliable way to distinguish between them.

Comment (THenry14) [reply crates/cairo-coverage/src/data_loader/loaded_data.rs:50]: Hmm, ok thanks! Let's leave it then, I will try to look into this further in profiler and assess on my own. Will keep you updated :+1:

**Review by THenry14 — APPROVED** [review]

File: crates/cairo-coverage/src/data_loader/loaded_data.rs (hunk 1 above)
Comment (ksew1) [reply crates/cairo-coverage/src/data_loader/loaded_data.rs:50]: `Snforge` already compiles to CASM using USC, so we could serialize it and pass it. Then, we wouldn't need to compile it, but we'll need to change the interface, and this implies that using this as a standalone binary won't be easy.

The same can be done in the profiler, so if we ever want to do that optimization, we can change this also there.

Comment (piotmag769) [reply crates/cairo-coverage/src/data_loader/loaded_data.rs:50]: It is not worth the effort, the gains would be really minimal (especially after https://github.com/software-mansion/universal-sierra-compiler/pull/57)

File: crates/cairo-coverage/src/data_loader/sierra_program.rs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+use cairo_lang_sierra::debug_info::DebugInfo;
+use cairo_lang_sierra::program::{Program, ProgramArtifact, VersionedProgram};
+use cairo_lang_sierra_to_casm::compiler::{CairoProgramDebugInfo, SierraToCasmConfig};
+use cairo_lang_sierra_to_casm::metadata::{calc_metadata, MetadataComputationConfig};
+use cairo_lang_starknet_classes::casm_contract_class::CasmContractClass;
+use cairo_lang_starknet_classes::contract_class::ContractClass;
+use serde::Deserialize;
+
+#[derive(Deserialize)]
+#[serde(untagged)]
```
Comment (piotmag769) [inline crates/cairo-coverage/src/data_loader/sierra_program.rs:11]: This is really nice

**Pull request merged**